File: tests/class/virtual-dispatch.swift
public class Foo {
	public func virtualCall(passthrough: Bool) -> String {
		return "Foo"
	}
	public final func staticCall() -> String {
		return "Static"
	}
}

public class Bar: Foo  {
	public override func virtualCall(passthrough: Bool) -> String {
		if (passthrough) {
			return super.virtualCall(passthrough: passthrough)
		} else {
			return "Bar"
		}
	}
	public final func newMethod() -> String {
		return "New method called on Bar"
	}
}

public func makeVirtualCall(onFoo foo: Foo, passthrough: Bool) -> String {
	return "Result: " + foo.virtualCall(passthrough: passthrough)
}

public func makeVirtualCall(onBar bar: Bar, passthrough: Bool) -> String {
	return "Result: " + bar.virtualCall(passthrough: passthrough)
}

public func makeStaticCall(onFoo foo: Foo) -> String {
	return "Result: " + foo.staticCall()
}

public func makeStaticCall(onBar bar: Bar) -> String {
	return "Result: " + bar.staticCall()
}

public func isBar(foo: Foo) -> Bool {
	return foo is Bar
}

public func makeCastAndCall(onFoo foo: Foo) -> String {
	if let bar = foo as? Bar {
		return bar.newMethod()
	} else {
		return "Not a Bar"
	}
}

File: tests/class/virtual-dispatch.js
export function makeVirtualCall$onFoo$passthrough$(foo, passthrough) {
  return "Result: " + foo.virtualCall$passthrough$(passthrough);
}
export function makeVirtualCall$onBar$passthrough$(bar, passthrough) {
  return "Result: " + bar.virtualCall$passthrough$(passthrough);
}
export function makeStaticCall$onFoo$(foo) {
  return "Result: Static";
}
export function makeStaticCall$onBar$(bar) {
  return "Result: Static";
}
export function isBar$foo$(foo) {
  return foo instanceof Bar;
}
export class Foo {
  virtualCall$passthrough$(passthrough) {
    return "Foo";
  }

}
export class Bar extends Foo {
  virtualCall$passthrough$(passthrough) {
    if (passthrough) {
      return super.virtualCall$passthrough$(passthrough);
    } else {
      return "Bar";
    }
  }

}